File: code/go/ch21/src/luago/state/api_access.go
package state

import "fmt"
import . "luago/api"

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_rawlen
func (self *luaState) RawLen(idx int) uint {
	val := self.stack.get(idx)
	switch x := val.(type) {
	case string:
		return uint(len(x))
	case *luaTable:
		return uint(x.len())
	default:
		return 0
	}
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_typename
func (self *luaState) TypeName(tp LuaType) string {
	switch tp {
	case LUA_TNONE:
		return "no value"
	case LUA_TNIL:
		return "nil"
	case LUA_TBOOLEAN:
		return "boolean"
	case LUA_TNUMBER:
		return "number"
	case LUA_TSTRING:
		return "string"
	case LUA_TTABLE:
		return "table"
	case LUA_TFUNCTION:
		return "function"
	case LUA_TTHREAD:
		return "thread"
	default:
		return "userdata"
	}
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_type
func (self *luaState) Type(idx int) LuaType {
	if self.stack.isValid(idx) {
		val := self.stack.get(idx)
		return typeOf(val)
	}
	return LUA_TNONE
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_isnone
func (self *luaState) IsNone(idx int) bool {
	return self.Type(idx) == LUA_TNONE
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_isnil
func (self *luaState) IsNil(idx int) bool {
	return self.Type(idx) == LUA_TNIL
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_isnoneornil
func (self *luaState) IsNoneOrNil(idx int) bool {
	return self.Type(idx) <= LUA_TNIL
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_isboolean
func (self *luaState) IsBoolean(idx int) bool {
	return self.Type(idx) == LUA_TBOOLEAN
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_istable
func (self *luaState) IsTable(idx int) bool {
	return self.Type(idx) == LUA_TTABLE
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_isfunction
func (self *luaState) IsFunction(idx int) bool {
	return self.Type(idx) == LUA_TFUNCTION
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_isthread
func (self *luaState) IsThread(idx int) bool {
	return self.Type(idx) == LUA_TTHREAD
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_isstring
func (self *luaState) IsString(idx int) bool {
	t := self.Type(idx)
	return t == LUA_TSTRING || t == LUA_TNUMBER
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_isnumber
func (self *luaState) IsNumber(idx int) bool {
	_, ok := self.ToNumberX(idx)
	return ok
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_isinteger
func (self *luaState) IsInteger(idx int) bool {
	val := self.stack.get(idx)
	_, ok := val.(int64)
	return ok
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_iscfunction
func (self *luaState) IsGoFunction(idx int) bool {
	val := self.stack.get(idx)
	if c, ok := val.(*closure); ok {
		return c.goFunc != nil
	}
	return false
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_toboolean
func (self *luaState) ToBoolean(idx int) bool {
	val := self.stack.get(idx)
	return convertToBoolean(val)
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_tointeger
func (self *luaState) ToInteger(idx int) int64 {
	i, _ := self.ToIntegerX(idx)
	return i
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_tointegerx
func (self *luaState) ToIntegerX(idx int) (int64, bool) {
	val := self.stack.get(idx)
	return convertToInteger(val)
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_tonumber
func (self *luaState) ToNumber(idx int) float64 {
	n, _ := self.ToNumberX(idx)
	return n
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_tonumberx
func (self *luaState) ToNumberX(idx int) (float64, bool) {
	val := self.stack.get(idx)
	return convertToFloat(val)
}

// [-0, +0, m]
// http://www.lua.org/manual/5.3/manual.html#lua_tostring
func (self *luaState) ToString(idx int) string {
	s, _ := self.ToStringX(idx)
	return s
}

func (self *luaState) ToStringX(idx int) (string, bool) {
	val := self.stack.get(idx)

	switch x := val.(type) {
	case string:
		return x, true
	case int64, float64:
		s := fmt.Sprintf("%v", x) // todo
		self.stack.set(idx, s)
		return s, true
	default:
		return "", false
	}
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_tocfunction
func (self *luaState) ToGoFunction(idx int) GoFunction {
	val := self.stack.get(idx)
	if c, ok := val.(*closure); ok {
		return c.goFunc
	}
	return nil
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_tothread
func (self *luaState) ToThread(idx int) LuaState {
	val := self.stack.get(idx)
	if val != nil {
		if ls, ok := val.(*luaState); ok {
			return ls
		}
	}
	return nil
}

// [-0, +0, –]
// http://www.lua.org/manual/5.3/manual.html#lua_topointer
func (self *luaState) ToPointer(idx int) interface{} {
	// todo
	return self.stack.get(idx)
}

File: code/csharp/ch13/luavm/state/ApiAccess.cs
using System;
using luavm.api;
using LuaType = System.Int32;

namespace luavm.state
{
    public partial class LuaState
    {
        public string TypeName(LuaType tp)
        {
            switch (tp)
            {
                case Constant.LUA_TNONE: return "no value";
                case Constant.LUA_TNIL: return "nil";
                case Constant.LUA_TBOOLEAN: return "boolean";
                case Constant.LUA_TNUMBER: return "number";
                case Constant.LUA_TSTRING: return "string";
                case Constant.LUA_TTABLE: return "table";
                case Constant.LUA_TFUNCTION: return "function";
                case Constant.LUA_TTHREAD: return "thread";
                default: return "userdata";
            }
        }

        public LuaType Type(int idx)
        {
            if (!_stack.IsValid(idx))
            {
                return Constant.LUA_TNONE;
            }

            var val = _stack.Get(idx);
            return LuaValue.TypeOf(val);
        }

        public bool IsNone(int idx)
        {
            return Type(idx) == Constant.LUA_TNONE;
        }

        public bool IsNil(int idx)
        {
            return Type(idx) == Constant.LUA_TNIL;
        }

        public bool IsNoneOrNil(int idx)
        {
            return Type(idx) <= Constant.LUA_TNIL;
        }

        public bool IsBoolean(int idx)
        {
            return Type(idx) == Constant.LUA_TBOOLEAN;
        }

        public bool IsString(int idx)
        {
            var t = Type(idx);
            return t == Constant.LUA_TSTRING || t == Constant.LUA_TNUMBER;
        }

        public bool IsNumber(int idx)
        {
            return ToNumberX(idx).Item2;
        }

        public bool IsInteger(int idx)
        {
            var val = _stack.Get(idx);
            return val is long;
        }

        public bool IsCsharpFunction(int idx)
        {
            var val = _stack.Get(idx);
            if (val is Closure c)
            {
                return c.CsharpFunc != null;
            }

            return false;
        }

        public bool ToBoolean(int idx)
        {
            var val = _stack.Get(idx);
            return ConvertToBoolean(val);
        }


        private static bool ConvertToBoolean(object val)
        {
            if (val == null)
            {
                return false;
            }

            switch (val)
            {
                case bool b: return b;
                default: return true;
            }
        }

        public double ToNumber(int idx)
        {
            return ToNumberX(idx).Item1;
        }

        public (double, bool) ToNumberX(int idx)
        {
            var val = _stack.Get(idx);
            return LuaValue.ConvertToFloat(val);
        }

        public (long, bool) ToIntegerX(int idx)
        {
            var val = _stack.Get(idx);
            return LuaValue.ConvertToInteger(val);
        }

        public long ToInteger(int idx)
        {
            var val = ToIntegerX(idx);
            return val.Item1;
        }

        public string ToString(int idx)
        {
            return ToStringX(idx).Item1;
        }

        public (string, bool) ToStringX(int idx)
        {
            var val = _stack.Get(idx);
            switch (val)
            {
                case string valStr: return (valStr, true);
                case long _:
                case double _:
                    var s = val;
                    _stack.Set(idx, s);
                    return (Convert.ToString(s), true);
                default: return (null, false);
            }
        }

        public CsharpFunction ToCsharpFunction(int idx)
        {
            var val = _stack.Get(idx);
            if (val is Closure c)
            {
                return c.CsharpFunc;
            }

            return null;
        }
    }
}